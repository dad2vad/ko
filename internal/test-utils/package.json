{
  "name": "@kosko/test-utils",
  "version": "0.0.0",
  "private": true,
  "engines": {
    "node": ">=14.18.0"
  },
  "main": "dist/index.node.cjs",
  "module": "dist/index.node.mjs",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "build-scripts"
  },
  "exports": {
    "types": "./dist/index.d.ts",
    "deno": "./dist/index.deno.mjs",
    "node": {
      "import": "./dist/index.node.mjs",
      "require": "./dist/index.node.cjs"
    },
    "import": "./dist/index.base.mjs",
    "require": "./dist/index.node.cjs"
  },
  "dependencies": {
    "execa": "^5.1.1",
    "temp-dir": "^2.0.0",
    "tmp": "^0.2.1",
    "tmp-promise": "^3.0.3"
  },
  "devDependencies": {
    "@kosko/build-scripts": "workspace:^",
    "@types/tmp": "^0.2.3"
  }
}
