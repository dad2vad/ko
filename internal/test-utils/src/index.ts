export * from "./getRejectedValue";
export * from "./installPackage";
export * from "./makeTempDir";
export * from "./makeTempFile";
export * from "./projectRoot";
