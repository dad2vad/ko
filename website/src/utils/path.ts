export const sep = "/";

export function isAbsolute(path: string): boolean {
  return path.startsWith(sep);
}

export function isRelative(path: string): boolean {
  return path.startsWith(`.${sep}`) || path.startsWith(`../${sep}`);
}

export function dirname(path: string): string {
  const index = path.lastIndexOf(sep);
  if (index === -1) return ".";
  return path.substring(0, index) || sep;
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf(sep);
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split(sep).filter(Boolean);
  const toParts = to.split(sep).filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join(sep);
}
