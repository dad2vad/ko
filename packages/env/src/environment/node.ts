import { getRequireExtensions } from "@kosko/require";
import { join } from "node:path";
import { formatPath } from "../paths";
import { toArray } from "@kosko/common-utils";
import {
  createEnvironment,
  createLoaderReducers,
  createReducerList
} from "./base";
import { Environment, Loader, ReducerExecutor, ReducerList } from "./types";
import { cwd as getCwd } from "node:process";

export function getNodeExtensions(): string[] {
  return getRequireExtensions().map((ext) => ext.substring(1));
}

/**
 * Options for creating a Node.js environment.
 *
 * @public
 */
export interface NodeEnvironmentOptions {
  /**
   * Current working directory (CWD).
   *
   * @defaultValue `process.cwd()`
   */
  cwd?: string;
}

export interface InternalNodeEnvironmentOptions extends NodeEnvironmentOptions {
  createReducerExecutor(this: void, reducers: ReducerList): ReducerExecutor;
  requireModule(this: void, environment: Environment, id: string): any;
  mergeValues(this: void, data: any[]): any;
}

export function createNodeEnvironment({
  cwd = getCwd(),
  createReducerExecutor,
  requireModule,
  mergeValues
}: InternalNodeEnvironmentOptions): Environment {
  function requireAllEnvs(template: string, component?: string) {
    if (!environment.env) return [];

    const envs = toArray(environment.env);

    return envs.map((env) => {
      const path = formatPath(template, {
        environment: env,
        ...(component && { component })
      });

      return requireModule(environment, join(environment.cwd, path));
    });
  }

  const loader: Loader = {
    global: () => requireAllEnvs(environment.paths.global),
    component: (name) => requireAllEnvs(environment.paths.component, name)
  };
  const reducers = createReducerList(createLoaderReducers(loader, mergeValues));
  const { reduce } = createReducerExecutor(reducers);
  const environment = createEnvironment({
    cwd,
    extensions: getNodeExtensions(),
    setReducers: reducers.setReducers,
    resetReducers: reducers.resetReducers,
    reduce
  });

  return environment;
}
