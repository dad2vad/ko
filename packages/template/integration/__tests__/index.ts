import execa from "execa";
import { copyFile, readFile } from "node:fs/promises";
import { join } from "node:path";
import { installPackage, makeTempDir, TempDir } from "@kosko/test-utils";

let args: string[];
let result: execa.ExecaReturnValue;
let options: execa.Options;
let tmpDir: TempDir;

beforeEach(async () => {
  tmpDir = await makeTempDir();

  const src = join(__dirname, "..", "bin.js");
  const dst = join(tmpDir.path, "bin.js");
  await copyFile(src, dst);
  await installPackage(tmpDir.path, "template");

  result = await execa(dst, args, {
    ...options,
    cwd: tmpDir.path,
    env: {
      LC_ALL: "en_US"
    }
  });
});

afterEach(async () => {
  await tmpDir.cleanup();
});

async function readFiles(cwd: string): Promise<string[]> {
  return Promise.all(
    ["foo", "bar/baz"].map((path) => readFile(join(cwd, path), "utf8"))
  );
}

describe("given args --help", () => {
  beforeAll(() => {
    args = ["--help"];
    options = {};
  });

  test("should return status code 0", () => {
    expect(result.exitCode).toEqual(0);
  });

  test("should print help", () => {
    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when required arg is not given", () => {
  beforeAll(() => {
    args = [];
    options = { reject: false };
  });

  test("should return status code 1", () => {
    expect(result.exitCode).toEqual(1);
  });

  test("should print error", () => {
    expect(result.stderr).toMatchSnapshot();
  });
});

describe("when cwd is not set", () => {
  beforeAll(() => {
    args = ["--foo", "bar", "--bar", "46.93"];
    options = {};
  });

  test("should return status code 0", () => {
    expect(result.exitCode).toEqual(0);
  });

  test("should write files to process.cwd", async () => {
    await expect(readFiles(tmpDir.path)).resolves.toMatchSnapshot();
  });
});

describe("when cwd is set", () => {
  let cwd: string;

  beforeAll(() => {
    cwd = join(tmpDir.path, "foo");
    args = ["--foo", "bar", "--bar", "46.93", "--cwd", cwd];
    options = {};
  });

  test("should return status code 0", () => {
    expect(result.exitCode).toEqual(0);
  });

  test("should write files to specified path", async () => {
    await expect(readFiles(cwd)).resolves.toMatchSnapshot();
  });
});
