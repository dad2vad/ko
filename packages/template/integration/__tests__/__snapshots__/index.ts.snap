// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given args --help should print help 1`] = `
"This is a fake template.

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
  --cwd      Path of working directory                   [string] [default: CWD]
  --foo      option foo                                                 [string]
  --bar      option bar                                      [number] [required]"
`;

exports[`when cwd is not set should write files to process.cwd 1`] = `
[
  "bar",
  "46.93",
]
`;

exports[`when cwd is set should write files to specified path 1`] = `
[
  "bar",
  "46.93",
]
`;

exports[`when required arg is not given should print error 1`] = `
"This is a fake template.

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
  --cwd      Path of working directory                   [string] [default: CWD]
  --foo      option foo                                                 [string]
  --bar      option bar                                      [number] [required]

Missing required argument: bar"
`;
